src/lua/engine_lua.c

```diff
@@ -0,0 +1,392 @@
+
+    /* Duplicate the function with __server__err__hanler and
+     * __redis__err_handler name for backwards compatibility. */
+    lua_pushstring(lua, REGISTRY_ERROR_HANDLER_NAME);
+    lua_gettable(lua, LUA_REGISTRYINDEX);
+    lua_setglobal(lua, "__server__err__handler");
+    lua_getglobal(lua, "__server__err__handler");
+    lua_setglobal(lua, "__redis__err__handler");
+}
+
+static void initializeLuaState(luaEngineCtx *lua_engine_ctx,
+                               subsystemType type) {
+    lua_State *lua = lua_open();
+
+    if (type == VMSE_EVAL) {
+        lua_engine_ctx->eval_lua = lua;
+    } else {
+        serverAssert(type == VMSE_FUNCTION);
+        lua_engine_ctx->function_lua = lua;
+    }
+
+    luaRegisterServerAPI(lua);
+    luaStateInstallErrorHandler(lua);
+
+    if (type == VMSE_EVAL) {
+        initializeEvalLuaState(lua);
+        luaStateLockGlobalTable(lua);
+    } else {
+        luaStateLockGlobalTable(lua);
+        luaFunctionInitializeLuaState(lua);
+    }
+}
+
+static struct luaEngineCtx *createEngineContext(void) {
+    luaEngineCtx *lua_engine_ctx = zmalloc(sizeof(*lua_engine_ctx));
+
+    initializeLuaState(lua_engine_ctx, VMSE_EVAL);
+    initializeLuaState(lua_engine_ctx, VMSE_FUNCTION);
+
+    return lua_engine_ctx;
+}
+
+static engineMemoryInfo luaEngineGetMemoryInfo(ValkeyModuleCtx *module_ctx,
+                                               engineCtx *engine_ctx,
+                                               subsystemType type) {
+    /* The lua engine is implemented in the core, and not in a Valkey Module */
+    serverAssert(module_ctx == NULL);
+
+    luaEngineCtx *lua_engine_ctx = engine_ctx;
+    engineMemoryInfo mem_info = {0};
+
+    if (type == VMSE_EVAL || type == VMSE_ALL) {
+        mem_info.used_memory += luaMemory(lua_engine_ctx->eval_lua);
+    }
+    if (type == VMSE_FUNCTION || type == VMSE_ALL) {
+        mem_info.used_memory += luaMemory(lua_engine_ctx->function_lua);
+    }
+
+    mem_info.engine_memory_overhead = zmalloc_size(engine_ctx);
+
+    return mem_info;
+}
+
+static compiledFunction **luaEngineCompileCode(ValkeyModuleCtx *module_ctx,
+                                               engineCtx *engine_ctx,
+                                               subsystemType type,
+                                               const char *code,
+                                               size_t timeout,
+                                               size_t *out_num_compiled_functions,
+                                               robj **err) {
+    /* The lua engine is implemented in the core, and not in a Valkey Module */
+    serverAssert(module_ctx == NULL);
+
+    luaEngineCtx *lua_engine_ctx = (luaEngineCtx *)engine_ctx;
+    compiledFunction **functions = NULL;
+
+    if (type == VMSE_EVAL) {
+        lua_State *lua = lua_engine_ctx->eval_lua;
+
+        if (luaL_loadbuffer(
+                lua, code, strlen(code), "@user_script")) {
```

Comment:
this line break looks odd.
```suggestion
        if (luaL_loadbuffer(lua, code, strlen(code), "@user_script")) {
```